### ISSUE TITLE
kj/filesystem-disk tests fail

### ISSUE THREAD

**Issue by dwrensha** [sandstorm-io/capnproto#469]
I'm seeing three test failures in the Ekam build on my Arch Linux box:

```
test: kj/filesystem-disk-old-kernel-test
test: kj/filesystem-disk-generic-test
  full log: tmp/kj/filesystem-disk-generic-test.log
  [ FAIL ] filesystem-disk-test.c++:201: DiskDirectory
test: kj/filesystem-disk-test
  full log: tmp/kj/filesystem-disk-test.log
  [ FAIL ] filesystem-disk-test.c++:201: DiskDirectory
```

```
$ cat tmp/kj/filesystem-disk-old-kernel-test.log 
[ TEST ] filesystem-disk-test.c++:98: DiskFile
[ PASS ] filesystem-disk-test.c++:98: DiskFile
[ TEST ] filesystem-disk-test.c++:176: DiskFile::copy()
[ PASS ] filesystem-disk-test.c++:176: DiskFile::copy()
[ TEST ] filesystem-disk-test.c++:201: DiskDirectory
kj/filesystem-disk-test.c++:259: failed: expected list[0] == "bar"
stack: 0x42fd2a 0x4212bc 0x4359dd 0x439167 0x434844 0x43467d 0x442a92 0x439167 0x4410e4 0x433bf2 0x7fd98ee53511 0x4045fa
kj/filesystem-disk-test.c++:260: failed: expected list[1] == "foo"
stack: 0x42fd2a 0x421322 0x4359dd 0x439167 0x434844 0x43467d 0x442a92 0x439167 0x4410e4 0x433bf2 0x7fd98ee53511 0x4045fa
kj/filesystem-disk-test.c++:266: failed: expected list[0].name == "bar"
stack: 0x42ff0a 0x4213eb 0x4359dd 0x439167 0x434844 0x43467d 0x442a92 0x439167 0x4410e4 0x433bf2 0x7fd98ee53511 0x4045fa
kj/filesystem-disk-test.c++:267: failed: expected list[0].type == FsNode::Type::DIRECTORY
stack: 0x43118a 0x42142a 0x4359dd 0x439167 0x434844 0x43467d 0x442a92 0x439167 0x4410e4 0x433bf2 0x7fd98ee53511 0x4045fa
kj/filesystem-disk-test.c++:268: failed: expected list[1].name == "foo"
stack: 0x42ff0a 0x421490 0x4359dd 0x439167 0x434844 0x43467d 0x442a92 0x439167 0x4410e4 0x433bf2 0x7fd98ee53511 0x4045fa
kj/filesystem-disk-test.c++:269: failed: expected list[1].type == FsNode::Type::FILE
stack: 0x4300ea 0x4214ce 0x4359dd 0x439167 0x434844 0x43467d 0x442a92 0x439167 0x4410e4 0x433bf2 0x7fd98ee53511 0x4045fa
[ FAIL ] filesystem-disk-test.c++:201: DiskDirectory
[ TEST ] filesystem-disk-test.c++:343: DiskDirectory symlinks
[ PASS ] filesystem-disk-test.c++:343: DiskDirectory symlinks
[ TEST ] filesystem-disk-test.c++:398: DiskDirectory link
[ PASS ] filesystem-disk-test.c++:398: DiskDirectory link
[ TEST ] filesystem-disk-test.c++:426: DiskDirectory copy
[ PASS ] filesystem-disk-test.c++:426: DiskDirectory copy
[ TEST ] filesystem-disk-test.c++:450: DiskDirectory move
[ PASS ] filesystem-disk-test.c++:450: DiskDirectory move
[ TEST ] filesystem-disk-test.c++:469: DiskDirectory createTemporary
[ PASS ] filesystem-disk-test.c++:469: DiskDirectory createTemporary
[ TEST ] filesystem-disk-test.c++:478: DiskDirectory replace directory with file
[ PASS ] filesystem-disk-test.c++:478: DiskDirectory replace directory with file
[ TEST ] filesystem-disk-test.c++:506: DiskDirectory replace file with directory
[ PASS ] filesystem-disk-test.c++:506: DiskDirectory replace file with directory
9 test(s) passed 
1 test(s) failed 
$ cat tmp/kj/filesystem-disk-generic-test.log 
[ TEST ] filesystem-disk-test.c++:98: DiskFile
[ PASS ] filesystem-disk-test.c++:98: DiskFile
[ TEST ] filesystem-disk-test.c++:176: DiskFile::copy()
[ PASS ] filesystem-disk-test.c++:176: DiskFile::copy()
[ TEST ] filesystem-disk-test.c++:201: DiskDirectory
kj/filesystem-disk-test.c++:259: failed: expected list[0] == "bar"
stack: 0x45552a 0x43fa7c 0x41428d 0x417a17 0x4130f4 0x412f2d 0x421342 0x417a17 0x41f994 0x4124a2 0x7fc654d27511 0x40432a
kj/filesystem-disk-test.c++:260: failed: expected list[1] == "foo"
stack: 0x45552a 0x43fae2 0x41428d 0x417a17 0x4130f4 0x412f2d 0x421342 0x417a17 0x41f994 0x4124a2 0x7fc654d27511 0x40432a
kj/filesystem-disk-test.c++:266: failed: expected list[0].name == "bar"
stack: 0x45570a 0x43fbab 0x41428d 0x417a17 0x4130f4 0x412f2d 0x421342 0x417a17 0x41f994 0x4124a2 0x7fc654d27511 0x40432a
kj/filesystem-disk-test.c++:267: failed: expected list[0].type == FsNode::Type::DIRECTORY
stack: 0x45698a 0x43fbea 0x41428d 0x417a17 0x4130f4 0x412f2d 0x421342 0x417a17 0x41f994 0x4124a2 0x7fc654d27511 0x40432a
kj/filesystem-disk-test.c++:268: failed: expected list[1].name == "foo"
stack: 0x45570a 0x43fc50 0x41428d 0x417a17 0x4130f4 0x412f2d 0x421342 0x417a17 0x41f994 0x4124a2 0x7fc654d27511 0x40432a
kj/filesystem-disk-test.c++:269: failed: expected list[1].type == FsNode::Type::FILE
stack: 0x4558ea 0x43fc8e 0x41428d 0x417a17 0x4130f4 0x412f2d 0x421342 0x417a17 0x41f994 0x4124a2 0x7fc654d27511 0x40432a
[ FAIL ] filesystem-disk-test.c++:201: DiskDirectory
[ TEST ] filesystem-disk-test.c++:343: DiskDirectory symlinks
[ PASS ] filesystem-disk-test.c++:343: DiskDirectory symlinks
[ TEST ] filesystem-disk-test.c++:398: DiskDirectory link
[ PASS ] filesystem-disk-test.c++:398: DiskDirectory link
[ TEST ] filesystem-disk-test.c++:426: DiskDirectory copy
[ PASS ] filesystem-disk-test.c++:426: DiskDirectory copy
[ TEST ] filesystem-disk-test.c++:450: DiskDirectory move
[ PASS ] filesystem-disk-test.c++:450: DiskDirectory move
[ TEST ] filesystem-disk-test.c++:469: DiskDirectory createTemporary
[ PASS ] filesystem-disk-test.c++:469: DiskDirectory createTemporary
[ TEST ] filesystem-disk-test.c++:478: DiskDirectory replace directory with file
[ PASS ] filesystem-disk-test.c++:478: DiskDirectory replace directory with file
[ TEST ] filesystem-disk-test.c++:506: DiskDirectory replace file with directory
[ PASS ] filesystem-disk-test.c++:506: DiskDirectory replace file with directory
9 test(s) passed 
1 test(s) failed 
$ cat tmp/kj/filesystem-disk-test.log 
[ TEST ] filesystem-disk-test.c++:98: DiskFile
[ PASS ] filesystem-disk-test.c++:98: DiskFile
[ TEST ] filesystem-disk-test.c++:176: DiskFile::copy()
[ PASS ] filesystem-disk-test.c++:176: DiskFile::copy()
[ TEST ] filesystem-disk-test.c++:201: DiskDirectory
kj/filesystem-disk-test.c++:259: failed: expected list[0] == "bar"
stack: 0x45a1fa 0x44a96c 0x42030d 0x423a97 0x41f174 0x41efad 0x42d3c2 0x423a97 0x42ba14 0x41e522 0x7fca40bcd511 0x4044ba
kj/filesystem-disk-test.c++:260: failed: expected list[1] == "foo"
stack: 0x45a1fa 0x44a9d2 0x42030d 0x423a97 0x41f174 0x41efad 0x42d3c2 0x423a97 0x42ba14 0x41e522 0x7fca40bcd511 0x4044ba
kj/filesystem-disk-test.c++:266: failed: expected list[0].name == "bar"
stack: 0x45a3da 0x44aa9b 0x42030d 0x423a97 0x41f174 0x41efad 0x42d3c2 0x423a97 0x42ba14 0x41e522 0x7fca40bcd511 0x4044ba
kj/filesystem-disk-test.c++:267: failed: expected list[0].type == FsNode::Type::DIRECTORY
stack: 0x45b65a 0x44aada 0x42030d 0x423a97 0x41f174 0x41efad 0x42d3c2 0x423a97 0x42ba14 0x41e522 0x7fca40bcd511 0x4044ba
kj/filesystem-disk-test.c++:268: failed: expected list[1].name == "foo"
stack: 0x45a3da 0x44ab40 0x42030d 0x423a97 0x41f174 0x41efad 0x42d3c2 0x423a97 0x42ba14 0x41e522 0x7fca40bcd511 0x4044ba
kj/filesystem-disk-test.c++:269: failed: expected list[1].type == FsNode::Type::FILE
stack: 0x45a5ba 0x44ab7e 0x42030d 0x423a97 0x41f174 0x41efad 0x42d3c2 0x423a97 0x42ba14 0x41e522 0x7fca40bcd511 0x4044ba
[ FAIL ] filesystem-disk-test.c++:201: DiskDirectory
[ TEST ] filesystem-disk-test.c++:343: DiskDirectory symlinks
[ PASS ] filesystem-disk-test.c++:343: DiskDirectory symlinks
[ TEST ] filesystem-disk-test.c++:398: DiskDirectory link
[ PASS ] filesystem-disk-test.c++:398: DiskDirectory link
[ TEST ] filesystem-disk-test.c++:426: DiskDirectory copy
[ PASS ] filesystem-disk-test.c++:426: DiskDirectory copy
[ TEST ] filesystem-disk-test.c++:450: DiskDirectory move
[ PASS ] filesystem-disk-test.c++:450: DiskDirectory move
[ TEST ] filesystem-disk-test.c++:469: DiskDirectory createTemporary
[ PASS ] filesystem-disk-test.c++:469: DiskDirectory createTemporary
[ TEST ] filesystem-disk-test.c++:478: DiskDirectory replace directory with file
[ PASS ] filesystem-disk-test.c++:478: DiskDirectory replace directory with file
[ TEST ] filesystem-disk-test.c++:506: DiskDirectory replace file with directory
[ PASS ] filesystem-disk-test.c++:506: DiskDirectory replace file with directory
[ TEST ] filesystem-disk-test.c++:537: DiskFile holes
[ PASS ] filesystem-disk-test.c++:537: DiskFile holes
10 test(s) passed 
1 test(s) failed 
```

**Comment by kentonv** [comment]
Interesting. I suspect the problem is that your filesystem returns its listing in a different order from mine. The test needs to be updated to be order-independent. Or, maybe listings should automatically be sorted -- not unreasonable considering that it's returned as one big array anyway.

Are you using a filesystem other than ext4?

**Comment by dwrensha** [comment]
The filesystem is ext3.

**Comment by kentonv** [comment]
Whoa, isn't ext3 like almost a decade obsolete?

**Comment by kentonv** [comment]
Fixed by #473
